### ISSUE TITLE
A NOT NULL predicate unexpectedly evaluates to TRUE

### ISSUE THREAD

**Issue by mrigger** [pingcap/tidb#16440]
Consider the following statements:
```sql
CREATE TABLE t1(c0 INT);
INSERT INTO t1(c0) VALUES (NULL);
SELECT t1.c0 FROM t1 WHERE NOT t1.c0; -- expected: {}, actual: {NULL}
```
Unexpectedly, the `SELECT` fetches a row, even though the `NOT t1.c0` should evaluate to `NULL`.

This seems to be a regression bug. On commit 72f457674238d28435b874af2899f7f684140a49, this still works correctly. On commit d26875db7e95b157fe57a5c68b8310f4212cbb32, this starts failing. Thus, I suspect that the fix for https://github.com/pingcap/tidb/pull/16108 is the culprit. /CC @XuHuaiyu 

I can reproduce this both on the latest master (d901863643afb1a40ae96b1ee800e8a8c6d5808a), and also checked that this bug still occurs with the https://github.com/pingcap/tidb/pull/16430 fix (commit 5370f815d812622fa4560bc546d3a5ce9d15e005).

**Comment by cfzjywxk** [comment]
@XuHuaiyu  PTAL

**Comment by XuHuaiyu** [comment]
This is a known problem, it should be fixed by https://github.com/pingcap/tidb/pull/15926

**Comment by mrigger** [comment]
Thanks! Okay, I'll consider this a duplicate bug report then. I'm looking forward to the PR getting merged, since this bug is blocking me from updating to the latest master.

**Comment by mrigger** [comment]
Any updates on this? :)

**Comment by XuHuaiyu** [comment]
@mrigger 
Already fixed in master/release-4.0/release-3.1/release-3.0.

related PRs:
https://github.com/pingcap/tidb/pull/16451
https://github.com/pingcap/tidb/pull/16542
